libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/DomainDiscoveryCoordinator.swift

```diff
@@ -0,0 +1,123 @@
+    ///
+    /// - Parameter action: The action to inspect.
+    /// - Returns: A `DomainDiscoveryResult` if the callback URL is detected and parsed; otherwise, `nil`.
+    ///
+    /// Call this from your WKNavigationDelegate when a navigation action occurs to detect and extract the result from the domain discovery callback URL.
+    @objc(handleWithWebAction:)
+    public func handle(action: WKNavigationAction) -> DomainDiscoveryResult? {
+        guard let url = action.request.url else {
+            return nil
+        }
+        if isDomainDiscoveryCallbackURL(url) {
+            let result = parseDiscoveryCallbackURL(url)
+            return result
+        }
+        return nil
+    }
+    
+    @objc
+    public func isDisoveryDomain(_ domain: String?, clientId: String?) -> Bool {
+        guard let domain = domain, let clientId = clientId else { return false }
+        let isDiscovery = domain.lowercased().contains(DomainDiscovery.URLComponent.path.rawValue)
+        let isSupportedClient = DomainDiscovery.supportedClientIds.contains(clientId)
+        return isDiscovery && isSupportedClient
+    }
+}
+
+extension DomainDiscoveryCoordinator {
+    private static func buildDiscoveryURL(clientId: String, clientVersion: String, domain: String, callbackURL: String = DomainDiscovery.callbackURL.rawValue) -> NSURL? {
+        var components = URLComponents()
+        components.scheme = DomainDiscovery.URLComponent.scheme.rawValue
+        components.host = domain.components(separatedBy: "/").first
+        components.path = DomainDiscovery.URLComponent.path.rawValue
+        components.queryItems = [
+            URLQueryItem(name: DomainDiscovery.URLComponent.QueryItem.clientID.rawValue, value: clientId),
+            URLQueryItem(name: DomainDiscovery.URLComponent.QueryItem.clientVersion.rawValue, value: clientVersion),
+            URLQueryItem(name: DomainDiscovery.URLComponent.QueryItem.callbackURL.rawValue, value: callbackURL)
+        ]
+        return components.url as NSURL?
+    }
+    
+    private func isDomainDiscoveryCallbackURL(_ url: URL?) -> Bool {
+        guard let url = url as URL? else { return false }
+        return url.absoluteString.lowercased().hasPrefix(DomainDiscovery.callbackURL.rawValue.lowercased())
+    }
+    
+    private func parseDiscoveryCallbackURL(_ url: URL?) -> DomainDiscoveryResult? {
+        guard let url = url as URL? else { return nil }
+        let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
+        let loginHint = components?.queryItems?.first(where: { $0.name == "login_hint" })?.value
+        var myDomain = components?.queryItems?.first(where: { $0.name == "my_domain" })?.value
+        if let domain = myDomain, domain.hasPrefix("https://") {
+            myDomain = String(domain.dropFirst("https://".count))
+        }
+        return DomainDiscoveryResult(loginHint: loginHint ?? "", myDomain: myDomain ?? "")
```

Comment:
if `myDomain` is nil, what happens with the empty string in the result? should the result be nil?

Reply:
Yeah, I think it coud just retun nil in that case instead of an empty value.  I'll update that.